### PR TITLE
DAOS-5866 test: Negative testing (OSA dmg command)

### PR DESCRIPTION
**Pull request by rpadma2**
Test-tag-hw-medium: pr,hw,medium,ib2 osa

Summary:
    - Pass invalid values to dmg pool exclude[reintegrate, drain] commands.
    - Check whether dmg provides appropriate errors.
    - Extend code is added but commented (due to DAOS-6838 issue)

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

### CODE REVIEW

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (shimizukko) [inline src/tests/ftest/osa/osa_dmg_negative_test.py:29]:
```suggestion
        # Dmg command test sequence
```

File: src/tests/ftest/osa/osa_dmg_negative_test.yaml
Comment (shimizukko) [inline src/tests/ftest/osa/osa_dmg_negative_test.yaml:58]: Just my preference, but you used test_seq in the code, so it would be easier to read if you use it here too.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (shimizukko) [inline src/tests/ftest/osa/osa_dmg_negative_test.py:81]: Can you call dmg system query every few sec to see if the additional server has come up instead of waiting for the fixed 5 sec?

**Review by shimizukko — APPROVED** [review]
LGTM

File: src/tests/ftest/util/apricot/apricot/test.py
Comment (rpadma2) [reply src/tests/ftest/util/apricot/apricot/test.py:1271]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.py:83]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.py:93]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.py:29]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.py:112]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.yaml
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.yaml:58]: Done.

File: src/tests/ftest/osa/osa_dmg_negative_test.py
Comment (rpadma2) [reply src/tests/ftest/osa/osa_dmg_negative_test.py]: Sleep not required... start_additional_servers should take care of handling that the server is up.

**Review by shimizukko — DISMISSED** [review]
LGTM

**Review by shimizukko — DISMISSED** [review]
LGTM

**Review by byron-marohn — APPROVED** [review]

**Comment by rpadma2** [comment]
Sylvia, The daoscore tests failures (timeout) has nothing to with this PR change. This PR add osa_dmg_negative_test case.. Not sure whether there is some existing failures. If not, I can create a ticket in this regard.

**Comment by rpadma2** [comment]
Created bug ticket https://jira.hpdd.intel.com/browse/DAOS-6843 for the DaosCoreTest failures for tracking purpose.

**Comment by rpadma2** [comment]
I am not seeing any failures on this run except issues archiving logs...

See these issues. It has nothing to do with the test runs.

23:48:38  Running /usr/lib/daos/TESTING/ftest/get_remote_files.sh -z -x -f /var/tmp/ftest/avocado/job-results/latest/daos_logs/*.log*
23:48:38  Error executing '/usr/lib/daos/TESTING/ftest/get_remote_files.sh -z -x -f /var/tmp/ftest/avocado/job-results/latest/daos_logs/*.log*':
23:48:38  	Output:
23:48:38  Logging get_remote_files.sh output to /tmp/get_remote_files_20210220_044828.log
23:48:38  Compressing files larger than 1M ...
23:48:38    Error detected compressing file

...
00:19:34  Running on wolf-[137-141]: set -eu; rc=0; copied=(); for file in /var/tmp/core.*; do if [ -e $file ]; then if [ ! -s $file ]; then ((rc++)); ls -al $file; else if sudo chmod 644 $file && scp $file wolf-137:/var/tmp/ftest/avocado/job-results/latest/stacktraces/${file##*/}-$(hostname -s); then copied+=($file); if ! sudo rm -fr $file; then ((rc++)); ls -al $file; fi; else ((rc++)); ls -al $file; fi; fi; fi; done; echo Copied ${copied[@]:-no files}; exit $rc
00:19:34      wolf-[137-141]: rc=0, output: Copied no files
00:19:34  ERROR: Detected one or more tests that failed archiving!

**Comment by rpadma2** [comment]
All tests passed except archiving issues.

**Comment by frostedcmos** [comment]
There are actually earlier errors in https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/daos/branches/PR-4633/runs/12/nodes/953/steps/1317/log/?start=0

[2021-02-20T04:04:54.062Z]  (1/1) ./osa/osa_dmg_negative_test.py:OSADmgNegativeTest.test_osa_dmg_cmd;properties-extra_servers-hosts-pool-0-1-test_sequence-90de:  PASS (168.62 s)
[2021-02-20T04:04:54.062Z] RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
....
[2021-02-20T04:04:54.638Z]       Running /usr/lib/daos/TESTING/ftest/cart/cart_logtest.py ...
[2021-02-20T04:04:54.638Z]         Error: details in /var/tmp/daos_testing/test_osa_dmg_cmd_daos_server_0.log_cart_testlog

If this is a negative testing you might have to modify some part of ftest to not perform log parsing for your particular test, or whitelist errors as log parsing seems to be failing here

**Comment by ashleypittman** [comment]
Indeed, the log file in question is https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4633/12/artifact/Functional_Hardware_Medium/osa/osa_dmg_negative_test.py/daos_logs/test_osa_dmg_cmd_daos_server_0.wolf-139.cart_logtest.log/*view*/ and the error is "ERROR: Invalid data in server.log on following line
02/20-04:03:26.34 wolf-139 DAOS[62860/62867] mgmt ERR  src/mgmt/srv_drpc.c:543 pool_change_target_state() Failed to set pool target up",

following that back to the src, we can see this line.

https://github.com/daos-stack/daos/blob/master/src/mgmt/srv_drpc.c#L542

It's printing a uuid_t with %s which is incorrect, I'm not sure why uuid_t is defined as a char array but it is.  The line in question should be changed to use the DF_UUID/DP_UUID macros.

**Comment by ashleypittman** [comment]
Furthermore looking at that file there look to be dozens of that class of bug, and a few cases where D_ASPRINTF() are used where D_STRNDUP would be more appropriate.

**Comment by sylviachanoiyee** [comment]
archiving issue might be related to DAOS-6541. could you trigger just the fn hw med tests again? i want to make sure the changes in apricot is not triggering the archiving issue even though issue may not be in this PR. thanks

**Review by ashleypittman — CHANGES_REQUESTED** [review]
> archiving issue might be related to DAOS-6541. could you trigger just the fn hw med tests again? i want to make sure the changes in apricot is not triggering the archiving issue even though issue may not be in this PR. thanks

The archiving issue is not DAOS-6541, the test has exposed an issue in the code that will need to be addressed before this can land.

**Review by byron-marohn — APPROVED** [review]

**Review by ashleypittman — COMMENTED** [review]
Sorry, my mistake, I only looked at the comments, the correct fix has been applied.

As I put there appear to be multiple errors of this type in this one file, I'll file a ticket for the wider problem.

**Review by shimizukko — DISMISSED** [review]
LGTM

**Review by sylviachanoiyee — CHANGES_REQUESTED** [review]
I just noticed that part of this test is skipped for DAOS-6838.
Can we split the test into smaller dmg test variant and use the SkipForTicket decorator to skip the non working test?
Using comments to skip makes tracking very difficult in JIRA.
Please create a ticket to re-enable the test as well and have it block DAOS-6838. Thanks!

**Comment by rpadma2** [comment]
Hi Sylvia, I am on vacation this week and will not be able to work on it till I come back. I created DAOS-6838 based on what I noticed during my testing. I will leave it to Brad and you decide on whether to land this PR and go forward or wait for my changes with skipForTicket...

**Comment by rpadma2** [comment]
Also Sylvia, I am not in favor splitting the test code. It will make the testing take longer for each test sequence and consume more CI test time.

**Comment by sylviachanoiyee** [comment]
> Also Sylvia, I am not in favor splitting the test code. It will make the testing take longer for each test sequence and consume more CI test time.

I think it should be ok to split it. The test now takes about 3 minutes. I see 2 advantages of splitting:
1. the non-working part can be skipped via skipForTicket.
2. makes moving test around between PR, master, weekly easier because we have more granularity

**Comment by rpadma2** [comment]
Split the test based on Sylvia's input. Added the skipForTicket pragma. The testing will use extend pool functionality on weekly testing...

**Review by byron-marohn — APPROVED** [review]

**Review by sylviachanoiyee — APPROVED** [review]

**Pull request merged**
